### ISSUE TITLE
Nano support

### ISSUE THREAD

**Issue by peci1** [botblox/botblox-manager-firmware#8]
You mentioned that Nano doesn't in the end need any special firmware and that it should work with the same one Switchblox uses. I successfully flashed Nano, but no settings seem to take effect (I can see that it successfully stores and loads the settings from the EEPROM, but the switch chip's behavior is without change).

**Comment by peci1** [comment]
Issuing this command, all ports should be isolated:

    python3.8 -m manager -D /dev/ttyUSB0 vlan --group 1 --group 2 --group 3 --group 4 --group 5

however, Nano still communicates without hesitation. I uploaded this command via the J-Link header-s RX pin and received confirmation on the test point TX pin, so the Nano definitely received and stored the configuration.

**Comment by AaronElijah** [comment]
Thanks Martin, I'm trying to replicate this now

**Comment by AaronElijah** [comment]
Update: I've managed to replicate the issue and even more concerning is that none of the commands (for VLAN or port mirroring) seem to be taking effect. This is surprising as I have managed to get this working on Nano before (the nanos I had used all had the VLAN configuration that I had previously set on them). I think this may be an issue with the way the source code is being built and flashed to the STM32L011F4 chip on the Nano.

I will keep posting if I have more information on the latter.

**Comment by peci1** [comment]
Ok, please, keep investigating. Just to be sure - which 3 ports are physically connected to the switch chip? Are they the same as ports 1, 2 and 3 on large Switchblox?

**Comment by AaronElijah** [comment]
Update:
- We've managed to program the Nano with firmware but we've found that there are several problematic issues.
1) First is that there is an issue with the bit banger being very unreliable. During test that when we reset the VLAN configuration (`botblox --device /dev/ttyUSB0 vlan --reset`), the effects didn't take effect until after a power cycle. 
2) Potentially an issue with the differences between STM32L011D4 and STM32L011F4 (MCU for Nano and SwitchBlox respectively). This is surprising as they are nominally the same MCU but with different pin shapes. However, when we rebuilt the project again but for STM32L011F4 the results appeared to be different (even though the `.bin` file flashed to the MCU should be exactly the same)...

Either way, I will upload the firmware that I used to successfully configure the Nano (either the source code or a hex file you can use to flash the MCU). May I ask what rev of Nano you are using? Is this something that could be problematic @joshelijah?

**Comment by AaronElijah** [comment]
On your second point - you are correct on both counts - the nano is a subset of the switchblox so to speak

**Comment by joshelijah** [comment]
Yes Martin will need to be using either Rev C or Rev D, which has the RX pin routed out onto the pogo pin header. Earlier revisions don't rout RX onto that pogo pin header.

**Comment by peci1** [comment]
Thanks for investigating. Ours is Rev C ;)

**Comment by peci1** [comment]
And it is actually possible to utilize the testpoints for accessing the TX pin, so I think it should also be possible to access the RX pin the same way on older revisions. It's just tricky to make solid contact. Maybe you could design a 3D-printed holder for the normal breadboard cables that would help with this? :)

**Comment by joshelijah** [comment]
Confirmed what Aaron said re the ports, mapping is below:

SwitchBlox Nano
IP175G          BOARD LABELLED PORT
0                    1
1                    2
2                    3

SwitchBlox Nano
IP175G          BOARD LABELLED PORT
0                    1
1                    2
2                    3
3                    4
4                    5

So yes, SwitchBlox Nano is a subset of the SwitchBlox mapping.

**Comment by joshelijah** [comment]
> And it is actually possible to utilize the testpoints for accessing the TX pin, so I think it should also be possible to access the RX pin the same way on older revisions. It's just tricky to make solid contact. Maybe you could design a 3D-printed holder for the normal breadboard cables that would help with this? :)

Yes you can absolutely use the test points, that's what we're using to do the debugging. To get easy access to them, we use PCBite probe holders, they are awesome and make my life so much easier for dealing with tiny hardware. Link below.

https://sensepeek.com/

**Comment by peci1** [comment]
>     1. First is that there is an issue with the bit banger being very unreliable. During test that when we reset the VLAN configuration (`botblox --device /dev/ttyUSB0 vlan --reset`), the effects didn't take effect until after a power cycle.

I thought this is kind of a (undesired?) property of the firmware. Even with large Switchblox, `--reset` has never worked for me without a dual reboot of the board and one more call of `--reset` after the first reboot. Have you tried with the erase command?

**Comment by AaronElijah** [comment]
My replication was similar, though I think it worked without an extra call of `--reset`(?). My feeling is that this might be a property of IP175G switch chip but I haven't proved that. What puzzles me about `--reset` is that it is essentially designed as a shorthand for writing the default register data the same way you would write custom register data so why it seems to give different results is odd. For example, there should be no behavioural difference with `botblox --device dev vlan --reset` vs `botblox --device dev vlan --group 1 2 3 4 5` (both are the default behaviour) and inspection
of the actual data sent via UART shows that only bits that don't correspond to any port actually differ, so there should be no difference in behavior. It could be that those bits, which I thought weren't doing anything actually are, in which case `reset` needs to be more complicated. 

When I was testing erase it seemed to work after a power cycle as well, which as a user I would expect as `erase` implies erasing the EEPROM. 

Come to think of it, it might be cleaner and more understandable to remove `--reset` for each command `vlan`, `mirror`, `tag-vlan`, etc, and use an `erase` command. Then a log message to the user asking them to power cycle the board? This way we just have one simple command for getting back the default behavior.

**Comment by peci1** [comment]
I think auto-rebooting the switch chip after each command would make things work clearer - the settings in the chip would always be in sync with what is in the EEPROM. That might save some of the surprises a user gets when playing with the CLI... So the process would be: 1) write (only) to EEPROM, 2) reboot the switch chip and let it read the config from EEPROM... i.e. there would be no direct reconfiguration of the chip while it is running, only at its startup

I can also imagine a different working of the `--reset` feature. Each subcommand could provide a list of MII registers it sets, and the firmware could walk the EEPROM and erase all commands working with these registers. So it would behave like a subcommand-wise erase. But it is based on the assumption that no two subcommands need to touch the same MII register (which holds for now).
